src/ess/dream/diagnostics.py

```diff
@@ -0,0 +1,228 @@
+
+        self._horizontal_selector_box = ipw.HBox([h_buttons])
+        self._vertical_selector_box = ipw.HBox([v_buttons])
+        super().__init__(
+            [
+                ipw.HBox([ipw.Label('X'), self._horizontal_selector_box]),
+                ipw.HBox([ipw.Label('Y'), self._vertical_selector_box]),
+            ]
+        )
+
+    def _make_buttons(
+        self, dims: tuple[str, ...], h_dim: str, v_dim: str
+    ) -> tuple[ipw.ToggleButtons, ipw.ToggleButtons]:
+        style = {'button_width': '10em'}
+        h_buttons = ipw.ToggleButtons(options=dims, value=h_dim, style=style)
+        v_buttons = ipw.ToggleButtons(options=dims, value=v_dim, style=style)
+        h_buttons.observe(self._observe_values, names='value')
+        v_buttons.observe(self._observe_values, names='value')
+        return h_buttons, v_buttons
+
+    @property
+    def _horizontal_buttons(self) -> ipw.ToggleButtons:
+        return self._horizontal_selector_box.children[0]
+
+    @property
+    def _vertical_buttons(self) -> ipw.ToggleButtons:
+        return self._vertical_selector_box.children[0]
+
+    def set_dims(self, new_dims: tuple[str, ...]) -> None:
+        self._horizontal_buttons.unobserve_all(name='value')
+        self._vertical_buttons.unobserve_all(name='value')
+        # self._forward_messages = False
+
+        old_h = self._horizontal_buttons.value
+        old_v = self._vertical_buttons.value
+        default_h, default_v = self._default_dims(new_dims)
+        new_h = old_h if old_h in new_dims else default_h
+        new_v = old_v if old_v in new_dims else default_v
+
+        h_buttons, v_buttons = self._make_buttons(new_dims, new_h, new_v)
+        self._horizontal_selector_box.children = [h_buttons]
+        self._vertical_selector_box.children = [v_buttons]
+
+    @staticmethod
+    def _default_dims(dims: tuple[str, ...]) -> tuple[str, str]:
+        return dims[math.ceil(len(dims) / 2)], dims[0]
+
+    @property
+    def value(self):
+        return {
+            'horizontal': self._horizontal_buttons.value,
+            'vertical': self._vertical_buttons.value,
+        }
+
+    def observe(self, handler, names: str | list[str], *args, **kwargs):
+        if names != 'value':
+            super().observe(handler, names, *args, **kwargs)
+        else:
+            self._value_observers.append(handler)
+
+    def _observe_values(self, change: dict) -> None:
+        clicked = change['owner']
+        other = (
+            self._vertical_buttons
+            if clicked is self._horizontal_buttons
+            else self._horizontal_buttons
+        )
+        if other.value == clicked.value:
+            other.value = change['old']
+        else:
+            # Only when dims are different to avoid unnecessary redraws
+            # and drawing with invalid dims.
+            for observer in self._value_observers:
+                observer(change)
+
+
+def _flat_voxel_figure(
```

Comment:
Does the flattening scheme only apply to dream, or could we make this a generic tool for other instruments like the MaskingTool in scippneutron?

Reply:
It can be generalised easily. Which other instruments have multi-logical-dimensional detectors?

Reply:
Most of them? I remember when Morten was talking about the EFU, all instruments seems to have some logical ordering. Loki for sure has banks, tubes, straws, pixels.

Reply:
I would suggest, we merge this here and ask for feedback. If that is positive, we can generalise the viewer and move it into ESSreduce. What do you think?